fdbserver/KeyValueStoreMemory.actor.cpp

```diff
@@ -504,56 +519,95 @@ class KeyValueStoreMemory final : public IKeyValueStore, NonCopyable {
 			    cipherKeys.cipherHeaderKey,
 			    getEncryptAuthTokenMode(EncryptAuthTokenMode::ENCRYPT_HEADER_AUTH_TOKEN_MODE_SINGLE),
 			    BlobCipherMetrics::KV_MEMORY);
-			BlobCipherEncryptHeader cipherHeader;
+			uint16_t headerSize;
 			Arena arena;
-			StringRef ciphertext =
-			    cipher.encrypt(plaintext, sizeof(int) + v1.size() + v2.size(), &cipherHeader, arena)->toStringRef();
-			log->push(StringRef((const uint8_t*)&cipherHeader, BlobCipherEncryptHeader::headerSize));
-			log->push(ciphertext);
+			if (CLIENT_KNOBS->ENABLE_CONFIGURABLE_ENCRYPTION) {
+				BlobCipherEncryptHeaderRef headerRef;
+				StringRef cipherText = cipher.encrypt(plaintext, v1.size() + v2.size(), &headerRef, arena);
+				Standalone<StringRef> headerRefStr = BlobCipherEncryptHeaderRef::toStringRef(headerRef);
+				headerSize = headerRefStr.size();
+				ASSERT(headerSize > 0);
+				log->push(StringRef((const uint8_t*)&headerSize, sizeof(headerSize)));
+				log->push(headerRefStr);
+				log->push(cipherText);
+			} else {
+				BlobCipherEncryptHeader cipherHeader;
+				StringRef ciphertext =
+				    cipher.encrypt(plaintext, v1.size() + v2.size(), &cipherHeader, arena)->toStringRef();
+				headerSize = BlobCipherEncryptHeader::headerSize;
+				ASSERT(headerSize > 0);
+				log->push(StringRef((const uint8_t*)&headerSize, sizeof(headerSize)));
+				log->push(StringRef((const uint8_t*)&cipherHeader, headerSize));
+				log->push(ciphertext);
+			}
 		}
 		return log->push("\x01"_sr); // Changes here should be reflected in OP_DISK_OVERHEAD
 	}
 
 	// In case the op data is not encrypted, simply read the operands and the zero fill flag.
 	// Otherwise, decrypt the op type and data.
 	ACTOR static Future<Standalone<StringRef>> readOpData(KeyValueStoreMemory* self,
-	                                                      OpHeader* h,
+	                                                      OpHeader h,
 	                                                      bool* isZeroFilled,
-	                                                      int* zeroFillSize) {
+	                                                      int* zeroFillSize,
+	                                                      bool encryptedOp) {
 		// Metadata op types to be excluded from encryption.
 		static std::unordered_set<OpType> metaOps = { OpSnapshotEnd, OpSnapshotAbort, OpCommit, OpRollback };
-		if (metaOps.count((OpType)h->op) == 0) {
+		if (metaOps.count((OpType)h.op) == 0) {
 			// It is not supported to open an encrypted store as unencrypted, or vice-versa.
-			ASSERT_EQ(h->op == OpEncrypted, self->enableEncryption);
+			ASSERT_EQ(encryptedOp, self->enableEncryption);
 		}
-		state int remainingBytes = h->len1 + h->len2 + 1;
-		if (h->op == OpEncrypted) {
+		// if encrypted op read the header size
+		state uint16_t headerSize = 0;
```

Comment:
`headerSize` is sizeof(BlobCipherEncryptXXX)`? If yes, please name variable as `encryptHeaderSize` for clarity. Thanks